x-pack/plugins/security/server/lib/authentication/providers/kerberos.ts

```diff
@@ -0,0 +1,280 @@
+    return '';
+  }
+
+  return authorization.split(/\s+/)[0].toLowerCase();
+}
+
+/**
+ * Provider that supports Kerberos request authentication.
+ */
+export class KerberosAuthenticationProvider extends BaseAuthenticationProvider {
+  /**
+   * Performs Kerberos request authentication.
+   * @param request Request instance.
+   * @param [state] Optional state object associated with the provider.
+   */
+  public async authenticate(request: Legacy.Request, state?: ProviderState | null) {
+    this.debug(`Trying to authenticate user request to ${request.url.path}.`);
+
+    const authenticationScheme = getRequestAuthenticationScheme(request);
+    if (authenticationScheme && authenticationScheme !== 'negotiate') {
+      this.debug(`Unsupported authentication scheme: ${authenticationScheme}`);
+      return AuthenticationResult.notHandled();
+    }
+
+    let authenticationResult = await this.authenticateViaHeader(request);
+    if (state && authenticationResult.notHandled()) {
+      authenticationResult = await this.authenticateViaState(request, state);
+      if (authenticationResult.failed() && isAccessTokenExpiredError(authenticationResult.error)) {
+        authenticationResult = AuthenticationResult.notHandled();
+      }
+    }
+
+    // If we couldn't authenticate by means of all methods above, let's try to check if Elasticsearch can
+    // start authentication mechanism negotiation, otherwise just return authentication result we have.
+    return authenticationResult.notHandled()
+      ? await this.authenticateViaSPNEGO(request, state)
+      : authenticationResult;
+  }
+
+  /**
+   * Invalidates access token retrieved in exchange for SPNEGO token if it exists.
+   * @param request Request instance.
+   * @param state State value previously stored by the provider.
+   */
+  public async deauthenticate(request: Legacy.Request, state?: ProviderState) {
+    this.debug(`Trying to deauthenticate user via ${request.url.path}.`);
+
+    if (!state || !state.accessToken) {
+      this.debug('There is no access token invalidate.');
+      return DeauthenticationResult.notHandled();
+    }
+
+    try {
+      const {
+        invalidated_tokens: invalidatedAccessTokensCount,
+      } = await this.options.client.callWithInternalUser('shield.deleteAccessToken', {
+        body: { token: state.accessToken },
+      });
+
+      if (invalidatedAccessTokensCount === 0) {
+        this.debug('User access token was already invalidated.');
+      } else if (invalidatedAccessTokensCount === 1) {
+        this.debug('User access token has been successfully invalidated.');
+      } else {
+        this.debug(
+          `${invalidatedAccessTokensCount} user access tokens were invalidated, this is unexpected.`
+        );
+      }
+    } catch (err) {
+      this.debug(`Failed invalidating user's access token: ${err.message}`);
+      return DeauthenticationResult.failed(err);
+    }
+
+    return DeauthenticationResult.redirectTo('/logged_out');
```

Comment:
__note:__ It seems `/logged_out` page would be appropriate here since with SPNEGO user will be automatically authenticated from any other page.

Reply:
Makes sense to me, we didn't have this when I was initially working on Kerberos, but I think it makes complete sense here.